Yang enumerations

dpino opened the pull request:
`compile_data_for_schema` (lib/yang/binary.lua) breaks if YANG data contains an enumeration. The problem is that the code expect enumerations to have a ctype. Enumerations should be treated as strings.

dpino committed:
Enumerations should be treated as strings when compiling to binary

dpino committed:
Add binary compilation of enumerations test

wingo commented:
Since you are working in this area, would you mind also adding a verifier for enumerated values, as we did for identityrefs?

wingo commented:
Since it's optional for enumerated values to have associated integer values, I guess it makes sense to represent enumerated values as strings.  It would be possible to represent enumerations where all values have an integer as int32 (see https://tools.ietf.org/html/rfc6020#section-9.6.4.2), which would allow for more efficient representation of lists containing enumerated values -- basically we could allow them to compile to the cdata representation.  But perhaps this isn't important and we can avoid doing that.  In summary I think going with strings is the right answer, yes.

wingo commented on src/lib/yang/binary.lua:
```diff
@@ -434,12 +437,30 @@ function selftest()
+            enum indetermiante {
+               value 2;
+            }
+            enum minor {
```
Can you leave off the value for one of these?  Just to make sure we are not requiring that the value be present.

wingo reviewed (commented):
Looks good, just one comment.

dpino committed:
Add enum initializator

dpino committed:
Add enum validator

dpino committed:
Remove one value attribute from enumerations definition

dpino commented:
wingo: Addressed comments and added an enum validator. I added an enum initializator as well to collect all the properties defined in https://tools.ietf.org/html/rfc7950#page-150. The validator only checks that a schema instance refers to an enumeration defined in the schema.

OTOH, according to RFC7950, an enumeration definition should comply several constraints (value max be a signed int 32 value, enumeration identities must not be duplicated). I can add these constraint validations in the schema. WDYT?

wingo commented on src/lib/yang/data.lua:
```diff
@@ -240,7 +240,15 @@ local function range_validator(range, f) return f end
-local function enum_validator(range, f) return f end
+local function enum_validator(enums, f)
+   if not enums or #enums == 0 then return f end
+   return function (val)
```
can we instead compute a table of valid values so that the lookup goes faster?  can be a followup

wingo commented on src/lib/yang/schema.lua:
```diff
@@ -404,6 +404,15 @@ local function init_range(node, loc, argument, children)
    node.reference = maybe_child_property(loc, children, 'reference', 'value')
 end
+local function init_enum(node, loc, argument, children)
+   node.value = require_argument(loc, argument)
```
can we call this "name" ?  "value" is confusing.  dunno

wingo commented on src/lib/yang/schema.lua:
```diff
@@ -404,6 +404,15 @@ local function init_range(node, loc, argument, children)
+local function init_enum(node, loc, argument, children)
+   node.value = require_argument(loc, argument)
+   local intvalue = maybe_child_property(loc, children, 'value', 'value')
+   if intvalue then node.intvalue = tonumber(intvalue) end
```
then this can be "value"

wingo reviewed (commented):
Looks good.  Consider the suggested changes; OK to fix in a followup.

dpino committed:
Address review comments

wingo commented on src/lib/yang/data.lua:
```diff
@@ -243,10 +243,10 @@ local function bit_validator(range, f) return f end
 local function enum_validator(enums, f)
    if not enums or #enums == 0 then return f end
```
i think this check will no longer work as enums is a string-keyed table instead of an array

dpino committed:
Fix enums validation

dpino commented:
@wingo right, the enums validation was not working as expected. I fixed it in the last commit.

wingo reviewed (approved):
great!
